## 01. Import libraries

In [2]:
# Import Libraries
import pandas as pd
import numpy as np
import os

In [4]:
# Turn Project Folder into Path
path = r'/Users/tanayawinder/Desktop/6-27-2024 Instant Cart Basket Analysis'

In [8]:
# Import merged pickle file

opm_derived = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','ords_prods_merged_derived.pkl'))

In [12]:
# Check output

opm_derived.shape

(32404859, 17)

In [16]:
opm_derived.head()

order_id  user_id  number_of_orders  order_day_of_week  order_hour_of_day  \
0   2539329        1                 1                  2                  8   
1   2398795        1                 2                  3                  7   
2    473747        1                 3                  3                 12   
3   2254736        1                 4                  4                  7   
4    431534        1                 5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_range  \
0         Soda        77              7     9.0  Mid range product   
1         Soda        77              7     9.0  Mid range product   
2         Soda        77              7     9.0  Mid range product   
3         Soda        77              7     9.0  Mid range product   
4         Soda        77              7     9.0  Mid range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy        Average orders  
1  Regularly busy    Slowest days        Average orders  
2  Regularly busy    Slowest days        Average orders  
3      Least busy    Slowest days        Average orders  
4      Least busy    Slowest days           Most orders

### Find aggregated mean of 'number_of_orders' col grouped by 'department_id' for entire dataset¶


In [21]:
# Groupby on ords_prods_merge 'department_id', aggregate mean # of orders

opm_derived.groupby('department_id').agg({'number_of_orders':['mean']})

number_of_orders
                          mean
department_id                 
1                    15.457838
2                    17.277920
3                    17.170395
4                    17.811403
5                    15.215751
6                    16.439806
7                    17.225802
8                    15.340650
9                    15.895474
10                   20.197148
11                   16.170638
12                   15.887671
13                   16.583536
14                   16.773669
15                   16.165037
16                   17.665606
17                   15.694469
18                   19.310397
19                   17.177343
20                   16.473447
21                   22.902379

A few things are different about the output of this aggregation on the entire dataset vs on the subset (in Exercise). Some of the values are different, such as Produce (4). In the subset it has a mean of 18.83, in the whole dataframe the mean is 17.81. We are also only presented the rows that appear on record in the subset (and as they appear, not in chronological order), whereas we are guaranteed all departments if aggregating the entire dataset.

### Create a loyalty flag for existing customers using the transform() and loc() functions

In [27]:
# Groupby on dataset by user_id, apply transform() on the “number_of_orders" col to generate max orders per user

opm_derived['max_order'] =  opm_derived.groupby(['user_id'])['number_of_orders'].transform(np.max)

/var/folders/tx/v4zzzx0x7735_gq8qjd1ddj00000gn/T/ipykernel_52589/3716214790.py:3: FutureWarning: The provided callable <function max at 0x109a042c0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  opm_derived['max_order'] =  opm_derived.groupby(['user_id'])['number_of_orders'].transform(np.max)


In [29]:
opm_derived.head()

order_id  user_id  number_of_orders  order_day_of_week  order_hour_of_day  \
0   2539329        1                 1                  2                  8   
1   2398795        1                 2                  3                  7   
2    473747        1                 3                  3                 12   
3   2254736        1                 4                  4                  7   
4    431534        1                 5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_range  \
0         Soda        77              7     9.0  Mid range product   
1         Soda        77              7     9.0  Mid range product   
2         Soda        77              7     9.0  Mid range product   
3         Soda        77              7     9.0  Mid range product   
4         Soda        77              7     9.0  Mid range product   

      busiest_day    busiest_days busiest_period_of_day  max_order  
0  Regularly busy  Regularly busy        Average orders         10  
1  Regularly busy    Slowest days        Average orders         10  
2  Regularly busy    Slowest days        Average orders         10  
3      Least busy    Slowest days        Average orders         10  
4      Least busy    Slowest days           Most orders         10

In [31]:
opm_derived.shape

(32404859, 18)

In [35]:
# Use new max_orders col to create loyalty flag

opm_derived.loc[opm_derived['max_order'] > 40, 'loyalty_flag'] =  'Loyal customer'

In [41]:
opm_derived.loc[(opm_derived['max_order'] <= 40) & (opm_derived['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [43]:
opm_derived.loc[opm_derived['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [47]:
# Check output

opm_derived.shape

(32404859, 19)

In [49]:
opm_derived.head()

order_id  user_id  number_of_orders  order_day_of_week  order_hour_of_day  \
0   2539329        1                 1                  2                  8   
1   2398795        1                 2                  3                  7   
2    473747        1                 3                  3                 12   
3   2254736        1                 4                  4                  7   
4    431534        1                 5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_range  \
0         Soda        77              7     9.0  Mid range product   
1         Soda        77              7     9.0  Mid range product   
2         Soda        77              7     9.0  Mid range product   
3         Soda        77              7     9.0  Mid range product   
4         Soda        77              7     9.0  Mid range product   

      busiest_day    busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy        Average orders         10   
1  Regularly busy    Slowest days        Average orders         10   
2  Regularly busy    Slowest days        Average orders         10   
3      Least busy    Slowest days        Average orders         10   
4      Least busy    Slowest days           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

In [51]:
opm_derived['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

### Determine whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [57]:
# Aggregate mean, min, max of prices of products by loyalty category

opm_derived.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

### Create 'type_of_spender' col to target different customers in marketing campaigns

In [62]:
# Create avg_spend col

opm_derived['avg_spend'] = opm_derived.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/tx/v4zzzx0x7735_gq8qjd1ddj00000gn/T/ipykernel_52589/2196032372.py:3: FutureWarning: The provided callable <function mean at 0x109a04cc0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  opm_derived['avg_spend'] = opm_derived.groupby(['user_id'])['prices'].transform(np.mean)


In [66]:
# Check output

opm_derived.shape

(32404859, 20)

In [68]:
opm_derived.head()

order_id  user_id  number_of_orders  order_day_of_week  order_hour_of_day  \
0   2539329        1                 1                  2                  8   
1   2398795        1                 2                  3                  7   
2    473747        1                 3                  3                 12   
3   2254736        1                 4                  4                  7   
4    431534        1                 5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_range  \
0         Soda        77              7     9.0  Mid range product   
1         Soda        77              7     9.0  Mid range product   
2         Soda        77              7     9.0  Mid range product   
3         Soda        77              7     9.0  Mid range product   
4         Soda        77              7     9.0  Mid range product   

      busiest_day    busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy        Average orders         10   
1  Regularly busy    Slowest days        Average orders         10   
2  Regularly busy    Slowest days        Average orders         10   
3      Least busy    Slowest days        Average orders         10   
4      Least busy    Slowest days           Most orders         10   

   loyalty_flag  avg_spend  
0  New customer   6.367797  
1  New customer   6.367797  
2  New customer   6.367797  
3  New customer   6.367797  
4  New customer   6.367797

In [72]:
# Create spender flag

opm_derived.loc[opm_derived['avg_spend']<10,'spender_flag']='Low_spender'

In [78]:
opm_derived.loc[opm_derived['avg_spend']>= 10, 'spender_flag'] = 'High_spender'

In [82]:
# Check spender flag output

opm_derived['spender_flag'].value_counts(dropna = False)

spender_flag
Low_spender     31770614
High_spender      634245
Name: count, dtype: int64

In [84]:
opm_derived[['user_id', 'avg_spend', 'spender_flag', 'prices']].head(60)

user_id  avg_spend spender_flag  prices
0         1   6.367797  Low_spender     9.0
1         1   6.367797  Low_spender     9.0
2         1   6.367797  Low_spender     9.0
3         1   6.367797  Low_spender     9.0
4         1   6.367797  Low_spender     9.0
5         1   6.367797  Low_spender     9.0
6         1   6.367797  Low_spender     9.0
7         1   6.367797  Low_spender     9.0
8         1   6.367797  Low_spender     9.0
9         1   6.367797  Low_spender     9.0
10       15   3.980556  Low_spender     9.0
11       15   3.980556  Low_spender     9.0
12       15   3.980556  Low_spender     9.0
13       15   3.980556  Low_spender     9.0
14       15   3.980556  Low_spender     9.0
15       19   8.372059  Low_spender     9.0
16       19   8.372059  Low_spender     9.0
17       19   8.372059  Low_spender     9.0
18       21   7.012683  Low_spender     9.0
19       31   8.137793  Low_spender     9.0
20       31   8.137793  Low_spender     9.0
21       43   7.444898  Low_spender     9.0
22       43   7.444898  Low_spender     9.0
23       52   6.356287  Low_spender     9.0
24       52   6.356287  Low_spender     9.0
25       52   6.356287  Low_spender     9.0
26       52   6.356287  Low_spender     9.0
27       52   6.356287  Low_spender     9.0
28       52   6.356287  Low_spender     9.0
29       52   6.356287  Low_spender     9.0
30       52   6.356287  Low_spender     9.0
31       52   6.356287  Low_spender     9.0
32       52   6.356287  Low_spender     9.0
33       52   6.356287  Low_spender     9.0
34       52   6.356287  Low_spender     9.0
35       52   6.356287  Low_spender     9.0
36       52   6.356287  Low_spender     9.0
37       67   7.656790  Low_spender     9.0
38       67   7.656790  Low_spender     9.0
39       67   7.656790  Low_spender     9.0
40       67   7.656790  Low_spender     9.0
41       67   7.656790  Low_spender     9.0
42       67   7.656790  Low_spender     9.0
43       67   7.656790  Low_spender     9.0
44       67   7.656790  Low_spender     9.0
45       67   7.656790  Low_spender     9.0
46       67   7.656790  Low_spender     9.0
47       67   7.656790  Low_spender     9.0
48       67   7.656790  Low_spender     9.0
49       67   7.656790  Low_spender     9.0
50       67   7.656790  Low_spender     9.0
51       67   7.656790  Low_spender     9.0
52       67   7.656790  Low_spender     9.0
53       67   7.656790  Low_spender     9.0
54       67   7.656790  Low_spender     9.0
55       67   7.656790  Low_spender     9.0
56       81   8.165385  Low_spender     9.0
57       81   8.165385  Low_spender     9.0
58       82   6.441221  Low_spender     9.0
59       82   6.441221  Low_spender     9.0

### Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column.¶

In [89]:
# Create a frequent user col, Groupyby user_id, aggreagate days_since_prior_order median

opm_derived['customer_frequency'] = opm_derived.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/tx/v4zzzx0x7735_gq8qjd1ddj00000gn/T/ipykernel_52589/2886579926.py:3: FutureWarning: The provided callable <function median at 0x109b41da0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  opm_derived['customer_frequency'] = opm_derived.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [93]:
# Check output

opm_derived.head(50)

order_id  user_id  number_of_orders  order_day_of_week  order_hour_of_day  \
0    2539329        1                 1                  2                  8   
1    2398795        1                 2                  3                  7   
2     473747        1                 3                  3                 12   
3    2254736        1                 4                  4                  7   
4     431534        1                 5                  4                 15   
5    3367565        1                 6                  2                  7   
6     550135        1                 7                  1                  9   
7    3108588        1                 8                  1                 14   
8    2295261        1                 9                  1                 16   
9    2550362        1                10                  4                  8   
10   2968173       15                15                  1                  9   
11   1870022       15                17                  2                 16   
12   1911383       15                18                  2                 11   
13   2715276       15                21                  1                  9   
14    487368       15                22                  1                 10   
15   2293453       19                 2                  5                 14   
16   1973799       19                 5                  6                 12   
17    532817       19                 7                  4                 17   
18   1573906       21                10                  3                 10   
19   1593000       31                10                  3                  8   
20   2231262       31                17                  3                 11   
21   2580647       43                 6                  4                 16   
22   2187180       43                 9                  4                 12   
23   2497897       52                 1                  1                  9   
24   1318871       52                 2                  1                 10   
25   1261384       52                 3                  1                 10   
26    580568       52                 5                  2                 10   
27   2428073       52                 7                  2                  7   
28    342306       52                 9                  1                 11   
29   1520353       52                10                  2                 10   
30   2542086       52                11                  2                  9   
31    944694       52                12                  1                 16   
32   1268191       52                15                  2                 11   
33   1122089       52                17                  2                 10   
34   1498922       52                18                  2                 10   
35    180919       52                19                  3                  8   
36   2695875       52                21                  5                 10   
37   3205960       67                 1                  3                 11   
38    970922       67                 2                  1                 15   
39    596585       67                 3                  1                 13   
40   2037087       67                 4                  2                  9   
41   2474111       67                 5                  3                 14   
42   1568481       67                 6                  3                 12   
43   2048527       67                 8                  2                 11   
44   2389031       67                 9                  2                 14   
45   2378562       67                10                  1                 10   
46   1515698       67                11                  1                 11   
47   1928050       67                12                  2                 12   
48   1949274       67              

In [95]:
opm_derived.shape

(32404859, 22)

In [101]:
# Create customer frequency flag

opm_derived.loc[opm_derived['customer_frequency'] > 20, 'frequency_flag'] = 'Non-frequent Customer'

In [103]:
opm_derived.loc[(opm_derived['customer_frequency']>10) & (opm_derived['customer_frequency']<=20),'frequency_flag'] = 'Regular customer'

In [105]:
opm_derived.loc[opm_derived['customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent Customer'

In [109]:
# Check flag output
opm_derived['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent Customer        21559853
Regular customer          7208564
Non-frequent Customer     3636437
nan                             5
Name: count, dtype: int64

In [113]:
opm_derived[['user_id', 'customer_frequency', 'frequency_flag']].head(50)

user_id  customer_frequency         frequency_flag
0         1                20.5  Non-frequent Customer
1         1                20.5  Non-frequent Customer
2         1                20.5  Non-frequent Customer
3         1                20.5  Non-frequent Customer
4         1                20.5  Non-frequent Customer
5         1                20.5  Non-frequent Customer
6         1                20.5  Non-frequent Customer
7         1                20.5  Non-frequent Customer
8         1                20.5  Non-frequent Customer
9         1                20.5  Non-frequent Customer
10       15                10.0      Frequent Customer
11       15                10.0      Frequent Customer
12       15                10.0      Frequent Customer
13       15                10.0      Frequent Customer
14       15                10.0      Frequent Customer
15       19                 6.0      Frequent Customer
16       19                 6.0      Frequent Customer
17       19                 6.0      Frequent Customer
18       21                 7.5      Frequent Customer
19       31                 6.0      Frequent Customer
20       31                 6.0      Frequent Customer
21       43                10.0      Frequent Customer
22       43                10.0      Frequent Customer
23       52                 8.0      Frequent Customer
24       52                 8.0      Frequent Customer
25       52                 8.0      Frequent Customer
26       52                 8.0      Frequent Customer
27       52                 8.0      Frequent Customer
28       52                 8.0      Frequent Customer
29       52                 8.0      Frequent Customer
30       52                 8.0      Frequent Customer
31       52                 8.0      Frequent Customer
32       52                 8.0      Frequent Customer
33       52                 8.0      Frequent Customer
34       52                 8.0      Frequent Customer
35       52                 8.0      Frequent Customer
36       52                 8.0      Frequent Customer
37       67                 7.0      Frequent Customer
38       67                 7.0      Frequent Customer
39       67                 7.0      Frequent Customer
40       67                 7.0      Frequent Customer
41       67                 7.0      Frequent Customer
42       67                 7.0      Frequent Customer
43       67                 7.0      Frequent Customer
44       67                 7.0      Frequent Customer
45       67                 7.0      Frequent Customer
46       67                 7.0      Frequent Customer
47       67                 7.0      Frequent Customer
48       67                 7.0      Frequent Customer
49       67                 7.0      Frequent Customer

### Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [116]:
# Export file as pickle

opm_derived.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'aggregated_ords_prods_merge.pkl'))